File: server/php/src/Presentation/Mappers/PaymentDetailsMapper.php
<?php

namespace OnlinePayments\ExampleApp\Presentation\Mappers;

use OnlinePayments\ExampleApp\Application\Domain\Payments\PaymentDetails\OperationOutput;
use OnlinePayments\ExampleApp\Application\Domain\Payments\PaymentDetails\OperationPaymentReferences;
use OnlinePayments\ExampleApp\Application\Domain\Payments\PaymentDetails\PaymentReferences;
use OnlinePayments\ExampleApp\Application\DTOs\GetPaymentDetails\ResponseDto;
use OnlinePayments\ExampleApp\Presentation\Models\GetPaymentDetails\Response;

final class PaymentDetailsMapper
{
    public static function mapResponse(ResponseDto $response): Response
    {
        $mapped = new Response();
        $mapped->id = $response->id;
        $mapped->status = $response->status;
        $mapped->statusOutput = $response->statusOutput;
        $mapped->operations = self::mapOperations($response);
        $mapped->paymentOutput = $response->paymentOutput;
        $mapped->hostedCheckoutSpecificOutput = $response->hostedCheckoutSpecificOutput;

        return $mapped;
    }

    private static function mapOperations(ResponseDto $response): ?array
    {
        if ($response->paymentOutput === null) return null;

        $operationOutput = new OperationOutput();
        $operationOutput->id = $response->id;
        $operationOutput->status = $response->status;
        $operationOutput->paymentMethod = $response->paymentOutput->paymentMethod;

        $operationOutput->amountOfMoney = $response->paymentOutput->amountOfMoney;

        $operationOutput->operationReferences = new OperationPaymentReferences();
        $operationOutput->operationReferences->merchantReference = $response->paymentOutput->references?->merchantReference ?? null;

        $operationOutput->references = new PaymentReferences();
        $operationOutput->references->merchantReference = $response->paymentOutput->references?->merchantReference ?? null;
        $operationOutput->references->merchantParameters = $response->paymentOutput->references?->merchantParameters ?? null;

        $operationOutput->statusOutput = $response->statusOutput;

        return [$operationOutput];
    }
}

File: server/ruby/app/presentation/mappers/payment_details/payment_details_mapper.rb
module Presentation
  module Mappers
    module PaymentDetailsMapper
      module_function

      def from_dto(dto)
        mapped = Presentation::Models::GetPaymentDetails::Response.new
        mapped.id = dto.id
        mapped.status = dto.status
        mapped.status_output = dto.status_output
        mapped.operations = build_operations(dto)
        mapped.payment_output = dto.payment_output
        mapped.hosted_checkout_specific_output = dto.hosted_checkout_specific_output
        mapped
      end

      def build_operations(dto)
        return nil if dto.payment_output.nil?

        operation_output = Business::Domain::Payments::PaymentDetails::OperationOutput.new
        operation_output.id = dto.id
        operation_output.status = dto.status
        operation_output.payment_method = dto.payment_output.payment_method
        operation_output.amount_of_money = dto.payment_output.amount_of_money

        operation_output.operation_references = Business::Domain::Payments::PaymentDetails::OperationPaymentReferences.new
        operation_output.operation_references.merchant_reference =
          dto.payment_output.references&.merchant_reference

        operation_output.references = Business::Domain::Payments::PaymentDetails::PaymentReferences.new
        operation_output.references.merchant_reference =
          dto.payment_output.references&.merchant_reference
        operation_output.references.merchant_parameters =
          dto.payment_output.references&.merchant_parameters

        operation_output.status_output = dto.status_output

        [operation_output]
      end

      private :build_operations
    end
  end
end